compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -37,144 +151,350 @@ pub struct MatchBranchSimplification;
+
+                // Otherwise we cannot optimize. Try another block.
+                _ => return false,
+            };
+            compare_types.push(compare_type);
+        }
 
-                    _ => unreachable!(),
+        for (other_val, other_target) in iter {
+            let other_stmts = &bbs[other_target].statements;
+            if compare_types.len() != other_stmts.len() {
+                return false;
+            }
+            for (f, s) in iter::zip(&compare_types, other_stmts) {
+                match (*f, &s.kind) {
+                    (CompareType::Same(f_s), s_s) if f_s == s_s => {}
+                    (
+                        CompareType::Eq(lhs_f, f_ty, val),
+                        StatementKind::Assign(box (lhs_s, Rvalue::Use(Operand::Constant(s_c)))),
+                    ) if lhs_f == lhs_s
+                        && s_c.const_.ty() == f_ty
+                        && s_c.const_.try_eval_scalar_int(tcx, param_env) == Some(val) => {}
+                    (
+                        CompareType::Discr(lhs_f, f_ty, is_signed),
+                        StatementKind::Assign(box (lhs_s, Rvalue::Use(Operand::Constant(s_c)))),
+                    ) if lhs_f == lhs_s && s_c.const_.ty() == f_ty => {
+                        let Some(f) = s_c.const_.try_eval_scalar_int(tcx, param_env) else {
+                            return false;
+                        };
+                        if is_signed
+                            && s_c.const_.ty().is_signed()
+                            && f.try_to_int(f.size()).unwrap()
+                                == ScalarInt::try_from_uint(other_val, discr_size)
+                                    .unwrap()
+                                    .try_to_int(discr_size)
+                                    .unwrap()
+                        {
+                            continue;
+                        }
+                        if Some(f) == ScalarInt::try_from_uint(other_val, f.size()) {
+                            continue;
+                        }
+                        return false;
+                    }
+                    _ => return false,
                 }
-            });
-
-            from.statements
-                .push(Statement { source_info, kind: StatementKind::StorageLive(discr_local) });
-            from.statements.push(Statement {
-                source_info,
-                kind: StatementKind::Assign(Box::new((
-                    Place::from(discr_local),
-                    Rvalue::Use(discr),
-                ))),
-            });
-            from.statements.extend(new_stmts);
-            from.statements
-                .push(Statement { source_info, kind: StatementKind::StorageDead(discr_local) });
-            from.terminator_mut().kind = first.terminator().kind.clone();
-            should_cleanup = true;
+            }
         }
+        self.transfrom_types = compare_types.into_iter().map(|c| c.into()).collect();
```

Comment:
Why not keep the whole `compare_types`?

Reply:
Due to `Same(&'a StatementKind<'tcx>)`, I don't have a good solution other than clone it.